envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoyListenersFactory.kt

```diff
@@ -0,0 +1,304 @@
+                )
+                .addFilterChains(createEgressFilterChain(group))
+                .build()
+
+        return listOf(ingressListener, egressListener)
+    }
+
+    private fun createIngressFilterChain(group: Group): FilterChain {
+        return FilterChain.newBuilder()
+                .addFilters(createIngressFilter(group))
+                .build()
+    }
+
+    private fun createEgressFilterChain(group: Group): FilterChain {
+        return FilterChain.newBuilder()
+                .addFilters(createEgressFilter(group))
+                .build()
+    }
+
+    private fun createEgressFilter(group: Group): Filter {
+        val connectionManagerBuilder = HttpConnectionManager.newBuilder()
+                .setStatPrefix("egress_http")
+                .setRds(egressRds(group.ads))
+                .setHttpProtocolOptions(egressHttp1ProtocolOptions())
+
+        return createFilter(connectionManagerBuilder, egressFilters, group, "egress")
+    }
+
+    private fun createFilter(
+        connectionManagerBuilder: HttpConnectionManager.Builder,
+        filters: List<HttpFilterFactory>,
+        group: Group,
+        accessLogType: String
+    ): Filter {
+        filters.forEach {
+            val filter = it(group)
+            if (filter != null) {
+                connectionManagerBuilder.addHttpFilters(filter)
+            }
+        }
+
+        // checked in EnvoyListenersFactory#createListeners
+        if (group.listenersConfig!!.accessLogEnabled) {
+            connectionManagerBuilder.addAccessLog(accessLog(group.listenersConfig!!.accessLogPath, accessLogType))
+        }
+
+        return Filter.newBuilder()
+                .setName("envoy.http_connection_manager")
+                .setTypedConfig(ProtobufAny.pack(
+                        connectionManagerBuilder.build()
+                ))
+                .build()
+    }
+
+    private fun egressHttp1ProtocolOptions(): Http1ProtocolOptions? {
+        return Http1ProtocolOptions.newBuilder()
+                .setAllowAbsoluteUrl(boolValue(true))
+                .build()
+    }
+
+    private fun egressRds(ads: Boolean): Rds {
+        val configSource = ConfigSource.newBuilder()
+                .setInitialFetchTimeout(durationInSeconds(egressRdsInitialFetchTimeout))
+
+        if (ads) {
+            configSource.setAds(AggregatedConfigSource.getDefaultInstance())
+        } else {
+            configSource.setApiConfigSource(defaultApiConfigSource)
+        }
+
+        return Rds.newBuilder()
+                .setRouteConfigName("default_routes")
+                .setConfigSource(
+                        configSource.build()
+                )
+                .build()
+    }
+
+    private fun createIngressFilter(group: Group): Filter {
+        val ingressHttp = HttpConnectionManager.newBuilder()
+                .setStatPrefix("ingress_http")
+                .setUseRemoteAddress(boolValue(group.listenersConfig!!.useRemoteAddress))
```

Comment:
Again, reaching to the raw structure which has been interpreted earlier to be valid at this point. This suggests you should introduce an intermediate representation that hints the validation has been done and no language specific force (`!!`) should come into play.